packages/react-data-grid/src/HeaderRow.tsx

```diff
@@ -0,0 +1,160 @@
+
+  shouldComponentUpdate(nextProps: Props) {
+    return (
+      nextProps.width !== this.props.width
+      || nextProps.height !== this.props.height
+      || nextProps.columns !== this.props.columns
+      || !shallowEqual(nextProps.style, this.props.style)
+      || this.props.sortColumn !== nextProps.sortColumn
+      || this.props.sortDirection !== nextProps.sortDirection
+    );
+  }
+
+  getHeaderCellType(column: Column): HeaderCellType {
+    if (column.filterable && this.props.filterable) {
+      return HeaderCellType.FILTERABLE;
+    }
+
+    if (column.sortable && this.props.rowType !== HeaderRowType.FILTER) {
+      return HeaderCellType.SORTABLE;
+    }
+
+    return HeaderCellType.NONE;
+  }
+
+  getFilterableHeaderCell(column: Column) {
+    const FilterRenderer = column.filterRenderer || FilterableHeaderCell;
+    return <FilterRenderer column={column} onChange={this.props.onFilterChange} />;
+  }
+
+  getSortableHeaderCell(column: Column) {
+    const sortDirection = this.props.sortColumn === column.key && this.props.sortDirection || DEFINE_SORT.NONE;
+    const sortDescendingFirst = column.sortDescendingFirst || false;
+    return (
+      <SortableHeaderCell
+        column={column}
+        rowType={this.props.rowType}
+        onSort={this.props.onSort}
+        sortDirection={sortDirection}
+        sortDescendingFirst={sortDescendingFirst}
+      />
+    );
+  }
+
+  getHeaderRenderer(column: Column) {
+    if (column.headerRenderer && !column.sortable && !this.props.filterable) {
+      return column.headerRenderer;
+    }
+    const headerCellType = this.getHeaderCellType(column);
+    switch (headerCellType) {
+      case HeaderCellType.SORTABLE:
+        return this.getSortableHeaderCell(column);
+      case HeaderCellType.FILTERABLE:
+        return this.getFilterableHeaderCell(column);
+      default:
+        return undefined;
+    }
+  }
+
+  getCells() {
+    const cells = [];
+    const frozenCells = [];
+    const { columns, rowType } = this.props;
+
+    for (let i = 0, len = getSize(columns); i < len; i++) {
+      const column = getColumn(columns, i);
+      const renderer = column.key === 'select-row' && rowType === HeaderRowType.FILTER ? <div /> : this.getHeaderRenderer(column);
+
+      const cell = (
+        <HeaderCell
+          ref={node => this.cells[i] = node}
+          column={column}
+          rowType={rowType}
+          height={this.props.height}
+          renderer={renderer}
+          onResize={this.props.onColumnResize}
+          onResizeEnd={this.props.onColumnResizeEnd}
+          onHeaderDrop={this.props.onHeaderDrop}
+          draggableHeaderCell={this.props.draggableHeaderCell}
+        />
+      );
+
+      if (isFrozen(column)) {
+        frozenCells.push(cell);
+      } else {
+        cells.push(cell);
+      }
+    }
+
+    return cells.concat(frozenCells);
+  }
+
+  setScrollLeft(scrollLeft: number): void {
+    this.props.columns.forEach((column?: Column, i?: number) => {
```

Comment:
Immutable's types aren't top notch, that's why we end up with awkward types here.
![image](https://user-images.githubusercontent.com/567105/56210657-38e8b080-604e-11e9-8daf-3bc3cfb44e56.png)
The package version is also 2 years old so that doesn't help.